chore: App router migration - general settings

joeauyeung opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

- Tackles the following settings categories
	- Profile
	- Security
	- Billing
	- Developer

- Adds missing pages in Admin, Billing and Developer sections
- Removes references to the pages routes
- Refactors how we handle the settings headers. Instead of relying on the `MetaContext` we now generate the settings header server side
    - In the `page.tsx` we get the session and t function server side.
    - We then use the `<SettingsHeader>` component to set the page title and description
    	- The `<SettingsHeader>` is abstracted from the `<SettingsLayout>`	

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

- For the new routes visit the normal pages routers and the`/future` routes

ibex088 committed:
added lockedSMS to future routes

ibex088 committed:
add orgMigrations routes to future

ibex088 committed:
correct metadata

ibex088 committed:
small fix

ibex088 committed:
add orgMigrations to future

joeauyeung committed:
Merge branch 'main' into app-router-migration/admin-settings

joeauyeung committed:
Move components to client components

joeauyeung committed:
Remove tRPC element from edit user RSC

joeauyeung committed:
Merge branch 'main' into app-router-migration/admin-settings

joeauyeung committed:
Get username for meta

joeauyeung committed:
Remove suspense query

joeauyeung committed:
Merge branch 'app-router-migration/admin-settings' of https://github.com/calcom/cal.com into app-router-migration/admin-settings

joeauyeung committed:
Remove orgMigrations from app router

joeauyeung committed:
Merge branch 'main' into app-router-migration/admin-settings

joeauyeung committed:
Type fix

joeauyeung committed:
Revert "Remove suspense query"

This reverts commit eadd814f6e4a5d6856d9218342b7909c22fe62c6.

joeauyeung committed:
Handle suspenseQuery in app router

joeauyeung committed:
User edit page, fetch data server side

joeauyeung committed:
Merge branch 'app-router-migration/admin-settings' of https://github.com/calcom/cal.com into app-router-migration/admin-settings

zomars committed:
Merge branch 'main' into app-router-migration/admin-settings

zomars committed:
Update yarn.lock

joeauyeung committed:
Merge branch 'main' into app-router-migration/admin-settings

joeauyeung committed:
Export getFixedT

joeauyeung committed:
Set PageWrapper as root layout for settings

joeauyeung committed:
Settings Layout accepts strings for shell heading

joeauyeung committed:
Add OOO to app router settings

joeauyeung committed:
Refactor layout for my-account pages

joeauyeung committed:
Remove instances of pages router from my-account

joeauyeung committed:
Refactor security pages

joeauyeung committed:
Add billing to app router

joeauyeung committed:
Add admin API link to layout

joeauyeung committed:
Add api keys page

joeauyeung committed:
Webhooks WIP

joeauyeung committed:
Refactor SettingsHeader to client component

joeauyeung committed:
Add webhook pages

joeauyeung committed:
Refactor API keys page

joeauyeung committed:
Add admin app page

joeauyeung committed:
Merge branch 'main' into app-router/settings-layouts

joeauyeung committed:
Type fix

hbjORbj committed:
fix types

hbjORbj commented:
Hey @joeauyeung ! I opened a [PR](https://github.com/calcom/cal.com/pull/16337), fixing type errors in this branch

hbjORbj committed:
fix developer/webhooks/[id] param value type error

hbjORbj committed:
remove unnecessary code

hbjORbj committed:
do not pass t prop to CreateNewWebhookButton

hbjORbj committed:
fix type errors in webhook-edit-view

hbjORbj committed:
fix the remaining type errors

anikdhabal commented:
> Hey @joeauyeung ! I opened a [PR](https://github.com/calcom/cal.com/pull/16337), fixing type errors in this branch

We need to fix the type error in this branch before merging, no need to create a new for that. @joeauyeung will fix the type error

hbjORbj commented:
@anikdhabal Hey! My PR is targeting this branch so I think it's not a problem! Also, I wanted to push directly to this branch but I've got no GH permissions yet. I think @joeauyeung already asked the team to give me a maintainer role, so it's all good.

hbjORbj commented:
Oh, I see you already closed the PR. That's ok, I will cherry-pick my commits and push to this branch once I get the GH permissions.

anikdhabal commented:
> Oh, I see you @anikdhabal already closed the PR. That's ok, I will cherry-pick my commits and push to this branch once I get the GH permissions.

Hey @hbjORbj, nice to meet you. Let's push the changes in this branch once you have permission

hbjORbj committed:
do not use prisma directly in generateMetadata

hbjORbj commented on apps/web/app/future/settings/admin/organizations/[id]/edit/page.tsx:
```diff
@@ -0,0 +1,54 @@
+    );
+  }
+
+  const teamName = await prisma.team.findFirst({
```
Thanks, addressed

hbjORbj committed:
Merge remote-tracking branch 'origin/main' into app-router/settings-layouts

hbjORbj committed:
remove use client if unnecessary

hbjORbj commented:
Hey @joeauyeung , I see that you introduce quite changes to some submodules in `/packages`; e.g., `/packages/trpc`. Can you leave some explanatory comments?

joeauyeung commented on packages/features/settings/appDir/SettingsHeader.tsx:
This is the big change in this folder. This is abstracted from the `SettingsLayout`. This way we can pass the page heading data directly to the server component rather that relying on `<Meta />` and `MetaContext`

joeauyeung committed:
Remove unused shell heading from SettingsLayoutAppDir

joeauyeung commented on packages/lib/server/repository/organization.ts:
We're aiming to slowly transition a lot of our logic into a repository pattern. For this PR I moved the logic from our `trpc` endpoints into the repository so it could be still be used in `trpc` and some server components if we need them.

joeauyeung reviewed (commented):
@hbjORbj some page components live in the `/packages` folder so that's why you're seeing a lot of changes there. I did comment on some changes I've added.

joeauyeung committed:
Merge branch 'main' into app-router/settings-layouts

hbjORbj committed:
Merge branch 'main' into app-router/settings-layouts

hbjORbj committed:
Merge branch 'main' into app-router/settings-layouts

hbjORbj committed:
Merge remote-tracking branch 'origin/main' into app-router/settings-layouts

hbjORbj committed:
improve metadata

hbjORbj committed:
fix billing page

hbjORbj committed:
Merge branch 'main' into app-router/settings-layouts

hbjORbj committed:
fix import in settings/teams

hbjORbj committed:
Merge branch 'main' into app-router/settings-layouts

hbjORbj commented on apps/web/app/future/settings/admin/organizations/[id]/edit/page.tsx:
```diff
@@ -0,0 +1,47 @@
+const Page = async ({ params }: { params: Params }) => {
+  const input = orgIdSchema.safeParse(params);
+
+  if (!input.success) redirect("/404");
```
I think we can replace `redirect("/404")` with `notFound()`.

hbjORbj replied:
I think we can replace `redirect("/404")` with `notFound()`.

joeauyeung committed:
Merge branch 'main' into app-router/settings-layouts

joeauyeung committed:
Merge branch 'app-router/settings-layouts' of https://github.com/calcom/cal.com into app-router/settings-layouts

joeauyeung committed:
Use next `notFound()`

hbjORbj committed:
Merge branch 'main' into app-router/settings-layouts

hbjORbj committed:
fix type check

hbjORbj committed:
fix type check

hbjORbj committed:
remove unused code

hbjORbj commented on apps/web/app/future/settings/my-account/calendars/page.tsx:
```diff
@@ -1,10 +1,41 @@
+      title={t("calendars")}
+      description={t("calendars_description")}
+      CTA={<AddCalendarButton />}>
+      <CalendarsView />
```
I think this should be `CalendarListContainer` instead? `CalendarsView` has the same CTA; e.g.,
```
const CalendarsView = () => {
  const { t } = useLocale();

  return (
    <>
      <Meta
        title={t("calendars")}
        description={t("calendars_description")}
        CTA={<AddCalendarButton />}
        borderInShellHeader={false}
      />
      <div className="mt-8">
        <CalendarListContainer />
      </div>
    </>
  );
};
```

hbjORbj reviewed (changes requested):
Hey, Joe, great work! I left a comment. Also,
1. `future/settings/my-account/calendars` has a compile error
2. `future/settings/admin/lockedSMS` has a double layout as in the screenshot. This is because the `getLayout` in `settings/layout.tsx` is applied to `settings/admin/lockedSMS` (which also has its own layout: admin layout). If you want to avoid this complication, you can go back to using WithLayout in each page.
3. Can you add feature flag env vars? 
<img width="1436" alt="Screenshot 2024-09-04 at 22 51 42" src="https://github.com/user-attachments/assets/d24ef4ad-6d23-48be-aab3-1ccb54b1adfc">

hbjORbj commented:
Also, if it helps, here's a list you can use to go through:

`settings/admin/lockedSMS` 
`settings/admin/organizations/[id]/edit` 
`settings/admin/users/[id]/edit` 
`settings/billing` 
`settings/developer/api-keys` 
`settings/developer/webhooks` 
`settings/developer/webhooks/new` 
`settings/developer/webhooks/[id]` 
`settings/my-account/appearance` 
`settings/my-account/calendars` 
`settings/my-account/conferencing` 
`settings/my-account/general` 
`settings/my-account/out-of-office` 
`settings/my-account/profile` 
`settings/organizations/admin-api` 
`settings/security/impersonation` 
`settings/security/password` 
`settings/security/sso` 
`settings/security/two-factor-auth`

hbjORbj committed:
Merge branch 'main' into app-router/settings-layouts

joeauyeung committed:
Fix calendar setting page

joeauyeung committed:
Separate settings pages into route groups

joeauyeung commented:
@hbjORbj the admin settings have a different layout than the other settings pages. What are your thoughts on having two route groups for the different layouts?

https://github.com/calcom/cal.com/pull/16312/commits/dcb2b6073fc50453e0620ea366a51623036aef2a

joeauyeung committed:
Merge branch 'main' into app-router/settings-layouts

joeauyeung committed:
Refactor admin settings pages

joeauyeung committed:
Merge branch 'app-router/settings-layouts' of https://github.com/calcom/cal.com into app-router/settings-layouts

joeauyeung committed:
Remove meta instance from billing page route

joeauyeung committed:
Update settings layoutAppDir

joeauyeung committed:
Refactor developer settings pages

joeauyeung committed:
Refactor out of office

joeauyeung committed:
Refactor my account settings

joeauyeung committed:
Refactor admin api page

joeauyeung committed:
Refactor security pages

joeauyeung committed:
Merge branch 'main' into app-router/settings-layouts

joeauyeung committed:
Type fix

joeauyeung commented:
@hbjORbj I also refactored the pages to remove instances of `<Meta />` from the app router pages. 

Two changes I made were:
- If the page was in the `/web/modules` folder and was just the `<Meta />` component and the page component. We can remove the page from the `/modules` and call the `<Meta />` component directly in the page route and import the page component directly in both
- In some skeleton loaders it would use the `<Meta />` component. This was used to set the meta on page load. However, since we can do this server side in the app router, I removed it from the skeleton loaders. I didn't see any noticeable differences between the two types of routers.

hbjORbj committed:
Merge remote-tracking branch 'origin/main' into app-router/settings-layouts

hbjORbj committed:
fix styling in settings layout

hbjORbj commented:
@joeauyeung Looks great to me! One issue was that `SettingsLayoutAppDir` was not synced with `SettingsLayout` in terms of styling, so I addressed it in this [commit](https://github.com/calcom/cal.com/pull/16312/commits/b44332a844d9f8ac3505b264d23589cd32d3c5de).

`settings/admin/lockedSMS` ✅
`settings/admin/organizations/[id]/edit` ✅
`settings/admin/users/[id]/edit` ✅
`settings/billing` ✅
`settings/developer/api-keys` ✅
`settings/developer/webhooks` ✅
`settings/developer/webhooks/new` ✅
`settings/developer/webhooks/[id]` ✅
`settings/my-account/appearance` ✅
`settings/my-account/calendars` ✅
`settings/my-account/conferencing` ✅
`settings/my-account/general` ✅
`settings/my-account/out-of-office` ✅
`settings/my-account/profile` ✅
`settings/organizations/admin-api` ✅
`settings/security/impersonation` ✅
`settings/security/password` ✅
`settings/security/sso` ✅
`settings/security/two-factor-auth` ✅

hbjORbj reviewed (approved):
@joeauyeung Looks great to me! One issue was that `SettingsLayoutAppDir` was not synced with `SettingsLayout` in terms of styling, so I addressed it in this [commit](https://github.com/calcom/cal.com/pull/16312/commits/b44332a844d9f8ac3505b264d23589cd32d3c5de).

`settings/admin/lockedSMS` ✅
`settings/admin/organizations/[id]/edit` ✅
`settings/admin/users/[id]/edit` ✅
`settings/billing` ✅
`settings/developer/api-keys` ✅
`settings/developer/webhooks` ✅
`settings/developer/webhooks/new` ✅
`settings/developer/webhooks/[id]` ✅
`settings/my-account/appearance` ✅
`settings/my-account/calendars` ✅
`settings/my-account/conferencing` ✅
`settings/my-account/general` ✅
`settings/my-account/out-of-office` ✅
`settings/my-account/profile` ✅
`settings/organizations/admin-api` ✅
`settings/security/impersonation` ✅
`settings/security/password` ✅
`settings/security/sso` ✅
`settings/security/two-factor-auth` ✅

hbjORbj commented:
@joeauyeung One UI issue I found is that settings pages of app router have a bottom border that doesn't exist in pages router

- pages router
<img width="820" alt="Screenshot 2024-09-08 at 21 01 22" src="https://github.com/user-attachments/assets/a8566779-3379-4ef8-b9b6-b9612957f0a1">


- app router
<img width="817" alt="Screenshot 2024-09-08 at 21 01 16" src="https://github.com/user-attachments/assets/7ff6b02a-cf85-4731-ae7f-db0568ab116a">
